core/src/main/scala/kafka/server/DelayedOperation.scala

```diff
@@ -307,22 +297,26 @@ class DelayedOperationPurgatory[T <: DelayedOperation](purgatoryName: String,
    */
   private class Watchers(val key: Any) {
 
-    private[this] val operations = new ConcurrentLinkedQueue[T]()
-    private[this] val approximateWatched = new AtomicInteger(0)
+    private case class WatchedOperation(op: T, purged: AtomicBoolean = new AtomicBoolean(false))
```

Comment:
@ijuma Yeah, that's fair. I did a little testing with `TestPurgatoryPerformance` and did notice some overall degradation (took a while for the results to stabilize, but the overall trend was clear). After thinking about it, maybe we could combine part of the old patch and synchronize only the purging of operations. That eliminates the possibility of concurrent removal, which was the main reason we needed `WatchedOperation` in the first place. What do you think?